File: components/brave_wallet_ui/common/hooks/token.ts
// Copyright (c) 2021 The Brave Authors. All rights reserved.
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this file,
// you can obtain one at https://mozilla.org/MPL/2.0/.

import * as React from 'react'
import {
  BraveWallet,
  GetBlockchainTokenInfoReturnInfo
} from '../../constants/types'

export default function useTokenInfo (
  getBlockchainTokenInfo: (address: string) => Promise<GetBlockchainTokenInfoReturnInfo>,
  visibleTokens: BraveWallet.BlockchainToken[],
  fullTokenList: BraveWallet.BlockchainToken[],
  selectedNetwork?: BraveWallet.NetworkInfo
) {
  const [tokenContractAddress, setTokenContractAddress] = React.useState<string>('')
  const [foundTokenInfoByContractAddress, setFoundTokenInfoByContractAddress] = React.useState<BraveWallet.BlockchainToken | undefined>()

  // Instead of having this be a useCallback hook here we are using useEffect to
  // handle the asynchronous getBlockchainTokenInfo fallback method.
  // That away each component that uses this hook will not have to handle
  // this async call individually.
  React.useEffect(() => {
    const contractAddress = tokenContractAddress.toLowerCase()
    if (contractAddress === '') {
      setFoundTokenInfoByContractAddress(undefined)
      return
    }

    const checkVisibleList = visibleTokens.find((token) => token.contractAddress.toLowerCase() === contractAddress)
    const checkedLists = checkVisibleList ?? (fullTokenList.find((token) => token.contractAddress.toLowerCase() === contractAddress))

    if (checkedLists) {
      setFoundTokenInfoByContractAddress(checkedLists)
      return
    }

    if (!checkedLists && selectedNetwork?.chainId === BraveWallet.MAINNET_CHAIN_ID) {
      getBlockchainTokenInfo(contractAddress).then((value: GetBlockchainTokenInfoReturnInfo) => {
        if (value.token) {
          setFoundTokenInfoByContractAddress(value.token)
          return
        }
        setFoundTokenInfoByContractAddress(undefined)
      }).catch(e => console.log(e))
    }
    setFoundTokenInfoByContractAddress(undefined)
  }, [tokenContractAddress, visibleTokens, fullTokenList, selectedNetwork?.chainId])
  return {
    onFindTokenInfoByContractAddress: setTokenContractAddress,
    foundTokenInfoByContractAddress
  }
}

File: components/brave_ads/core/internal/common/challenge_bypass_ristretto/token.cc
/* Copyright (c) 2022 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at https://mozilla.org/MPL/2.0/. */

#include "brave/components/brave_ads/core/internal/common/challenge_bypass_ristretto/token.h"

#include "brave/components/brave_ads/core/internal/common/challenge_bypass_ristretto/blinded_token.h"
#include "brave/components/brave_ads/core/internal/common/challenge_bypass_ristretto/challenge_bypass_ristretto_util.h"

namespace brave_ads::cbr {

namespace {

std::optional<challenge_bypass_ristretto::Token> Create(
    const std::string& token_base64) {
  if (token_base64.empty()) {
    return std::nullopt;
  }

  return ValueOrLogError(
      challenge_bypass_ristretto::Token::DecodeBase64(token_base64));
}

}  // namespace

Token::Token() : token_(challenge_bypass_ristretto::Token::Random()) {}

Token::Token(const std::string& token_base64) : token_(Create(token_base64)) {}

Token::Token(const Token& other) = default;

Token& Token::operator=(const Token& other) = default;

Token::Token(Token&& other) noexcept = default;

Token& Token::operator=(Token&& other) noexcept = default;

Token::~Token() = default;

bool Token::operator==(const Token& other) const {
  return EncodeBase64().value_or("") == other.EncodeBase64().value_or("");
}

bool Token::operator!=(const Token& other) const {
  return !(*this == other);
}

Token Token::DecodeBase64(const std::string& token_base64) {
  return Token(token_base64);
}

std::optional<std::string> Token::EncodeBase64() const {
  if (!token_ || !has_value()) {
    return std::nullopt;
  }

  return token_->EncodeBase64();
}

std::optional<BlindedToken> Token::Blind() {
  if (!token_ || !has_value()) {
    return std::nullopt;
  }

  return ValueOrLogError<challenge_bypass_ristretto::BlindedToken,
                         BlindedToken>(token_->Blind());
}

std::ostream& operator<<(std::ostream& os, const Token& token) {
  os << token.EncodeBase64().value_or("");
  return os;
}

}  // namespace brave_ads::cbr
